Bar105 beherith stuff: Add a new GameFramePost + more tracy

Beherith opened the pull request:
Please review, dont merge yet, as more stuff incoming.

Beherith committed:
Add a new GameFramePost(int frameNum) gadget callin, which is the very last callin when a SimFrame ends. Useful for batching sendtounsynced and other things.

Beherith committed:
Add Tracy zones to a few minor areas that were previously left out.

sprunk commented:
What's the use case for `GameFramePost`? Isn't `GameFramePost(n)` the same as `GameFrame(n+1)`?

Beherith committed:
Replace the spring_time.sleep(true) which always waits 15.6ms with a socket-select-timeout based approach.

Beherith committed:
Allow for Tracy Plots from Lua, new UnsyncedCtrl callouts: LuaTracyPlotConfig, LuaTracyPlot

LuaTracyPlotConfig(number plotIndex , string plotType, bool step, bool fill, number color)
LuaTracyPlot(number plotIndex, number plotValue)
Plot indices go from 1-9, I could not figure out how to pass a name to the plot. Any help there is welcome!

Beherith committed:
Add function ldocs to LuaTracyPlotConfig and LuaTracyPlot

Beherith committed:
Add LuaUnsyncedRead GetLastSwapBuffersEnd and GetLastSwapBuffersDuration

GetLastSwapBuffersEnd(bool micros) returns a timer
GetLastSwapBuffersDuration() returns a float in milliseconds

Beherith committed:
LuaUnsyncedRead::GetFrameTimer allow queries for high-resolution microsecond level timers

Beherith committed:
Further Clarify LuaTracyPlotConfig ldoc strings

Beherith committed:
Add a new Lua Callin: widget:UpdateTimeOffset, that runs as the first callin of every update-draw

This is needed to override the frameTimeOffset from Lua

Beherith committed:
TracyPlotConfig and TracyPlot now take plotName strings as first arguments

thanks @sprunk

sprunk committed:
Add a comment.

Beherith committed:
Document sources of hard-coded duration CameraTransition's

Beherith committed:
Allow overriding SMF squareTex mip levels via SMFMipLevelOverride configint for  testing

The new config int SMFMipLevelOverride has the following settings:
[0-3] hard coded mip level choice in LoadSquareTexture
[-1] old, dynamic mip level selection
[-2] old, dynamic mip level selection and forced aging of out-of-los squaretex to miplevel 3

Default is 0.

Beherith committed:
Config Bool ROAMBufferSubData to allow glBufferData instead of glBufferSubData in ROAM indexVBO uploads to prevent rendering pipeline stalls

Config Bool ROAMBufferSubData, default true
[true]: uses old, glBufferSubData call to upload changes, this can stall rendering pipeline.
[false] uses new, glBufferData to always allocate a new buffer, this may be suboptimal in memory usage

sprunk commented on docker-build/history.txt:
```diff
@@ -0,0 +1,80 @@
+    5  sudo su
+    6  su root
+    7  sudo ls
+    8  sudo nano /etc/default/locale
```
delete the file (not because it's chaff but because it reveals you use `nano` as the editor. we can't take humiliating public PR hits like that)

sprunk commented:
this PR seems to be about a handful of things very loosely related to each other, perhaps the ready parts can be split off?
* GameFramePost is a decent addition and can be added standalone
* tracy stuff looks nice and i think can be isolated as well?
* ROAM also seems like the others don't depend on it

Beherith committed:
Add further Tracy zones to shadowhandler and worlddrawer

Beherith committed:
DWMFlush: allow forcing of Windows Desktop Compositing to prevent rendering of frames that wont ever be displayed.

If vSync is on, and triple buffering is not active, and a frame runs over time of vblank, then the frame still gets rendered, but will never actually be displayed (gets dropped). This can be wasteful of resources.
One can check for dropped frames via nVidia's FrameView app.
The new configuration bool DWMFlush (default false/off) forces a DWM Flush before SwapWindow.
This setting may help for capturing the game with OBS.

lhog commented:
Yeah we need to distill it into separate PRs.

Beherith commented:
Yeah this entire branch is stupid together, i will split the different segments into their own prs. The only real arc around them is improving frame pacing. FYI there are 33 more files changed and wip in my stash, i just commited the stuff here that are in a good state.

Beherith commented:
Because this started as something that was supposed to be easy and simple, and went very deep very quickly.

Beherith committed:
Add new 'm' character to texture options, which disables trilinear mipmap filtering for both dds  and uncompressed textures.

Because the default GL_LINEAR_MIPMAP_LINEAR is 2 lookups, while GL_LINEAR_MIPMAP_NEAREST is 1 lookup,

sprunk reviewed (changes requested):
Should be distilled into small, focused changes

Beherith commented:
Indeed, I am slowly working on splitting these into smaller chunks. TY @sprunk

sprunk commented:
Is there anything left here? Most commits here have been put into nice focused PRs.

Beherith committed:
FINAL COMMIT TO STASH GARBAGE
